### PR TITLE
 fix(gatsby-plugin-page-creator): Find nested files for knownCollections

### PR DESCRIPTION
**Pull request by rogermparent**
## Description

This PR fixes a minor issue where filesystem collection routes whose pages are only located within a directory with the route path (e.g. `a/{b.slug}/index.tsx` and/or `a/{b.slug}/other.tsx` but not `a/{b.slug}.tsx`) are not recognized during the `globby` call that finds files `knownCollections`, which down the line prevents these collections from having the `setFieldsOnGraphQLNodeType` logic applied to them and the `gatsbyPath` field never appears on those collections.

### Documentation

One could probably consider the [nested routes](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#nested-routes) and [gatsbyPath](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#gatsbypath-example) sections of the File System Route API docs combined to cover the whole of this feature and imply the fix moves more toward intended behavior.

### Tests

~~Started off with and relied on manual testing to confirm the bug and that this fix works. On this version that just adds an option to the `globby` call, I didn't see an existing test nor add a new one since there's no logic in this repo added and I'm not sure how best to introduce faking files for globs.~~

~~I took another shot at this earlier by changing the glob to `**/**{**}**(/**)` so all nested files would be picked up, complete with a test using `micromatch` which `globby` uses at its core, but I feel that this version is better despite being less tested.~~

EDIT: Upon feedback, I've changed this PR to the earlier version described above, along with some suggested changes. There's now a unit test for the glob at `src/__tests__/path-utils.ts` Everything else still applies.

I believe the original error stems from a corner case of [`picomatch`'s globstar behavior](https://github.com/micromatch/picomatch#basic-globbing) (and `micromatch` and `globby` by extension), quoted here:

> Note that ** will only match path separators (/, and \\ on Windows) when they are the only characters in a path segment. Thus, foo**/bar is equivalent to foo*/bar, and foo/a**b/bar is equivalent to foo/a*b/bar, and more than two consecutive stars in a glob path segment are regarded as a single star. Thus, foo/***/bar is equivalent to foo/*/bar.

Since the current glob has the `**{**}**` as the last segment instead of `**` alone, the globstars don't expand for nested directories. Can't blame whoever did it, this seems incredibly easy to miss with the key docs being buried under three layers of packages.

## Related Issues

Couldn't find anything related, it's a bit of a niche use-case at the intersection of using the FS Route API + `gatsbyPath` + collections whose queries only appear in directories and not files.

### CODE REVIEW

File: packages/gatsby-plugin-page-creator/src/gatsby-node.ts
Comment (pieh) [inline packages/gatsby-plugin-page-creator/src/gatsby-node.ts:426]: This doesn't seem to work as we would expect. Upon some local testing this seems to return paths to directories that match without including paths to files under those directories: For `a/{b.slug}/index.tsx` case it would just return `a/{b.slug}` instead of actual path to the template file we expect.

It seems to me that is caused more by glob pattern, and we shouldn't use `onlyFiles: false`. For example changing pattern to:

```diff
-const pagesGlob = `**/**\\{*\\}**`
+const pagesGlob = [`**/**\\{*\\}**/**`, `**/**\\{*\\}**`];
```

made it return what I would expect which is `a/{b.slug}/index.tsx`. It seems like current pattern only match if final segment in path has curly braces. Adding `/**` to the pattern make it match when there was curly braces in previous segments, but drops previous behaviour so it seems like those 2 patterns need to be combined? Or maybe there is syntax to have 1 pattern, but we didn't find one.

Given potential problems this could introduce I would suggest adding some tests for this:
1. Move lines 424-426 to separately utility function ( https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-page-creator/src/path-utils.ts is good place for it)
2. add some unit tests - you can use fixtures directory for test inputs

**Review by pieh — CHANGES_REQUESTED** [review]
Current change seems not right, see inline comment

File: packages/gatsby-plugin-page-creator/src/gatsby-node.ts
Comment (rogermparent) [reply packages/gatsby-plugin-page-creator/src/gatsby-node.ts:426]: Thanks for the feedback! I'll bring back my other version which fits more inline with what is asked here, complete with unit test and all!

Comment (rogermparent) [reply packages/gatsby-plugin-page-creator/src/gatsby-node.ts]: > It seems like current pattern only match if final segment in path has curly braces. Adding /** to the pattern make it match when there was curly braces in previous segments, but drops previous behaviour so it seems like those 2 patterns need to be combined? Or maybe there is syntax to have 1 pattern, but we didn't find one.

From what I tried, you can only do it with two basic patterns or use `()` from the advanced glob features to make a pattern (`**/**{**}**(/**`)). I'm not sure which of the two would be preferable.

File: packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts
Comment (rogermparent) [inline packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts:16]: This unit test runs the pattern against a list using the library underlying `globby`- is this enough or do we want to mock the filesystem and use `globby` itself? My opinion is that this is good enough for the logic we want to test, but there's good arguments on either side so I'm happy to defer to maintainers' opinion.

File: packages/gatsby-plugin-page-creator/src/path-utils.ts
Comment (rogermparent) [inline packages/gatsby-plugin-page-creator/src/path-utils.ts:128]: Could be `**/**\\{*\\}**(/**)` or `['**/**\\{*\\}**/**','**/**\\{*\\}**']`, not sure which is preferable.

Comment (LekoArts) [reply packages/gatsby-plugin-page-creator/src/path-utils.ts:130]: With the comment above I think it's fine 👍

File: packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts
Comment (LekoArts) [reply packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts]: `globby` could in theory switch its implementation under the hood in the future (e.g. switching to a Rust NAPI implementation to speed things up) and then that test would be irrelevant but still pass.

So we'll want to test `findCollectionPageFiles` and mock out anything that's necessary :)

File: packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts
Comment (rogermparent) [inline packages/gatsby-plugin-page-creator/src/__tests__/path-utils.ts:20]: This new test uses a fixture to test `findCollectionPageFiles` directly. The fixture files are blank because their content doesn't matter for this purpose, but I could easily see it expanded to be used for other logic in this plugin.

**Comment by rogermparent** [comment]
Sorry for the request spam, I thought I errantly clicked a "remove request" but was unaware the behavior of the review request menu is to remove other requests if one is clicked.

**Review by pieh — APPROVED** [review]

**Comment by LekoArts** [comment]
Until `5.9.0` is released, you can use this fix with `gatsby@5.9.0-next.1`

**Pull request merged**
